File: src/tool/DevilDaggersInfo.Tool.DistributeApp/ApiHttpClient.fs
module DevilDaggersInfo.Tool.DistributeApp.ApiHttpClient

open System.IO
open System.Net
open System.Net.Http
open System.Net.Http.Headers
open System.Net.Http.Json
open System.Text.Json
open DevilDaggersInfo.Api.Admin.Tools
open DevilDaggersInfo.Api.Main.Authentication

let login =
    let client = new HttpClient()
    let loginRequest = new HttpRequestMessage(
        HttpMethod.Post,
        "https://devildaggers.info/api/authentication/login",
        Content = JsonContent.Create(JsonSerializer.Deserialize<LoginRequest>(File.ReadAllText("appsettings.json"))))
    let loginAsync = async {
        let! response = client.SendAsync(loginRequest) |> Async.AwaitTask
        match response.StatusCode with
        | HttpStatusCode.OK -> ()
        | _ -> failwith $"Unsuccessful status code from login '{response.StatusCode}'"

        let! loginResponse = response.Content.ReadFromJsonAsync<LoginResponse>() |> Async.AwaitTask
        return loginResponse
    }
    let loginResponse = loginAsync |> Async.RunSynchronously
    loginResponse

let upload projectName projectVersion toolBuildType toolPublishMethod outputZipFilePath loginToken =
    let distribution = AddDistribution(
        Name = projectName,
        Version = projectVersion,
        BuildType = toolBuildType,
        PublishMethod = toolPublishMethod,
        ZipFileContents = File.ReadAllBytes(outputZipFilePath),
        UpdateVersion = true,
        UpdateRequiredVersion = false)
    let jsonContent = JsonContent.Create(distribution)
    let mutable uploadRequest = new HttpRequestMessage(
        HttpMethod.Post,
        "https://devildaggers.info/api/admin/tools",
        Content = jsonContent)
    uploadRequest.Headers.Authorization <- AuthenticationHeaderValue("Bearer", loginToken)

    let client = new HttpClient()
    let uploadAsync = async {
        let! response = client.SendAsync(uploadRequest) |> Async.AwaitTask
        return response
    }
    let uploadResponse = uploadAsync |> Async.RunSynchronously

    match uploadResponse.StatusCode with
    | HttpStatusCode.OK -> ()
    | _ -> failwith $"Unsuccessful status code from upload '{uploadResponse.StatusCode}'"

File: src/dev-util/DevilDaggersInfo.DevUtil.DistributeApp/ApiHttpClient.cs
using DevilDaggersInfo.Api.Admin.Tools;
using DevilDaggersInfo.Api.Main.Authentication;
using System.Net;
using System.Net.Http.Json;
using System.Text.Json;

namespace DevilDaggersInfo.Tool.DistributeApp;

public static class ApiHttpClient
{
	public static async Task<LoginResponse> LoginAsync()
	{
		using HttpRequestMessage loginRequest = new(HttpMethod.Post, "https://devildaggers.info/api/authentication/login");
		loginRequest.Content = JsonContent.Create(JsonSerializer.Deserialize<LoginRequest>(await File.ReadAllTextAsync("appsettings.json")));

		using HttpClient client = new();
		HttpResponseMessage response = await client.SendAsync(loginRequest);
		if (response.StatusCode != HttpStatusCode.OK)
			throw new($"Unsuccessful status code from login '{response.StatusCode}'");

		return await response.Content.ReadFromJsonAsync<LoginResponse>() ?? throw new("Could not deserialize login response.");
	}

	public static async Task UploadAsync(string projectName, string projectVersion, ToolBuildType toolBuildType, ToolPublishMethod toolPublishMethod, string outputZipFilePath, string loginToken)
	{
		AddDistribution addDistribution = new()
		{
			Name = projectName,
			Version = projectVersion,
			BuildType = toolBuildType,
			PublishMethod = toolPublishMethod,
			ZipFileContents = await File.ReadAllBytesAsync(outputZipFilePath),
			UpdateVersion = true,
			UpdateRequiredVersion = false,
		};

		using HttpRequestMessage uploadRequest = new(HttpMethod.Post, "https://devildaggers.info/api/admin/tools");
		uploadRequest.Content = JsonContent.Create(addDistribution);
		uploadRequest.Headers.Authorization = new("Bearer", loginToken);

		using HttpClient client = new();
		HttpResponseMessage response = await client.SendAsync(uploadRequest);
		if (response.StatusCode != HttpStatusCode.OK)
			throw new($"Unsuccessful status code from upload '{response.StatusCode}'");
	}
}
